### PR TITLE
Remove arguments bound to a variable that has no value provided in a query

### PR DESCRIPTION
**Pull request by annapawlicka**
This PR fixes issue #20 
+ Remove arguments bound to a variable that has no value provided in a query
  + null values, e.g. `foo(bar: null)` are values and are not removed
+ Update test schema
  + new tests to test missing/null variables
  + changed resolver for `hero` query. Previously if nullable argument `episode` was not provided, the resolver would still provide a value even though `default-value` was not specified. This behavior would mask an error about non-nullable character (`hero`) being null. Quite a few tests had to be updated to add `episode` argument (their logic has not changed) but I think this new behavior makes more sense.

### CODE REVIEW

File: dev-resources/com/walmartlabs/test_schema.clj [hunk 1]
```diff
@@ -233,6 +242,43 @@
                             :resolve (fn [_ _ _]
                                        "<Single Value>")}}}
 
+    :villain
```
Comment (hlship) [inline dev-resources/com/walmartlabs/test_schema.clj]: I know this is out of the blue but ... I really think that putting more code into test_schema.clj is a problem; I've been adding lots of more narrowly focused tests and namespaces and I feel that is really the right way to go.

Comment (hlship) [inline dev-resources/com/walmartlabs/test_schema.clj]: For example, each change here forces a change to the introspection tests.

Comment (annapawlicka) [inline dev-resources/com/walmartlabs/test_schema.clj]: Working on moving new test data and tests to a new namespace.

**Comment by annapawlicka** [comment]
@hlship I removed new object/query from test_schema.clj and updated separate variable tests.

File: src/com/walmartlabs/lacinia/parser.clj [hunk 2]
```diff
@@ -629,7 +629,13 @@
         ;; This is kind of a juxt buried in a map. Each value is a function that accepts
         ;; the variables and returns the actual value to use.
         (fn [variables]
-          (map-vals #(% variables) dynamic-args))))))
+          (->> (map-vals #(% variables) dynamic-args)
+               (filter (fn [[variable value]]
+                         (let [[_ user-defined-var-name] (get arguments variable)]
+                           ;; keep arguments that have a matching variable provided
+                           (or (contains? variables user-defined-var-name)
+                               ;; or have a default-value specified
```
Comment (hlship) [inline src/com/walmartlabs/lacinia/parser.clj]: I think there's a tiny bit of ambiguity here between a default value of nil or the lack of a default value at all.  This code treats a default value of nil as a lack of default value.  However, I'm willing to let that slide until/unless that becomes an issue.

**Comment by annapawlicka** [comment]
Right, I can remove the changes I made to `hero` query too, considering we want to move away from using test-schema.

File: src/com/walmartlabs/lacinia/parser.clj (hunk 2 above)
Comment (annapawlicka) [inline src/com/walmartlabs/lacinia/parser.clj]: I added a test to catch this scenario. Will try to fix it in this PR.

**Comment by annapawlicka** [comment]
I'm closing this PR while I'm making more changes.

**Comment by annapawlicka** [comment]
This PR is now ready for a review, @hlship @bcarrell

File: src/com/walmartlabs/lacinia/internal_utils.clj [hunk 3]
```diff
@@ -57,6 +57,19 @@
         persistent!
         (with-meta (meta m)))))
 
+(defn filter-kvs
+  "Builds a new map passing each key and value to predicate."
```
Comment (hlship) [inline src/com/walmartlabs/lacinia/internal_utils.clj]: The predicate only seems to involve the `key`, not the `value`.

File: src/com/walmartlabs/lacinia/parser.clj [hunk 4]
```diff
@@ -629,7 +633,12 @@
         ;; This is kind of a juxt buried in a map. Each value is a function that accepts
         ;; the variables and returns the actual value to use.
         (fn [variables]
-          (map-vals #(% variables) dynamic-args))))))
+          (->> (map-vals #(% variables) dynamic-args)
+               ;; keep arguments that have a matching variable provided.
+               ;; :value in value-map might be NULL but it's still a
+               ;; provided value (e.g. may be used to indicate deletion)
+               (filter-kvs (fn [variable value-map] (some? value-map)))
+               (map-vals (fn [v] (:value v)))))))))
```
Comment (hlship) [inline src/com/walmartlabs/lacinia/parser.clj]: `(map-vals :value)` would work here

File: src/com/walmartlabs/lacinia/internal_utils.clj (hunk 3 above)
Comment (annapawlicka) [inline src/com/walmartlabs/lacinia/internal_utils.clj]: Yep. It didn't before I refactored some of the stuff. Will fix it.

Comment (hlship) [inline src/com/walmartlabs/lacinia/internal_utils.clj]: Actually, filter-keys is what this is doing!

**Comment by annapawlicka** [comment]
Ok, I think it's much simpler now @hlship

**Pull request merged**